config/src/main/java/com/typesafe/config/ConfigDocument.java

```diff
@@ -0,0 +1,40 @@
+package com.typesafe.config;
+
+/**
+ * An object parsed from the original input text, which can be used to
+ * replace individual values and exactly render the original text of the
+ * input.
+ *
+ * <p>
+ * Because this object is immutable, it is safe to use from multiple threads and
+ * there's no need for "defensive copies."
+ *
+ * <p>
+ * <em>Do not implement interface {@code ConfigNode}</em>; it should only be
+ * implemented by the config library. Arbitrary implementations will not work
+ * because the library internals assume a specific concrete implementation.
+ * Also, this interface is likely to grow new methods over time, so third-party
+ * implementations will break.
+ */
+public interface ConfigDocument {
+    /**
+     * Returns a new ConfigDocument that is a copy of the current ConfigDocument,
+     * but with the desired value set at the desired path. If the path exists, it will
+     * remove all duplicates before the final occurrence of the path, and replace the value
+     * at the final occurrence of the path. If the path does not exist, it will be added.
+     *
+     * @param path the path at which to set the desired value
+     * @param newValue the value to set at the desired path, represented as a string. This
+     *                 string will be parsed into a ConfigNode, and the text will be inserted
+     *                 as-is into the document, with leading and trailing whitespace removed.
```

Comment:
I've written a test and verified the behavior. 

One thing to note: when parsing a single JSON value for node replacement, if a concatenation is given, the parser will simply parse the first value in the concatenation into a node and ignore all other items in the concatenation. Is that acceptable? Or would you prefer it error out when a concatenation is given and JSON is being parsed?

Also, do you want me to modify the `setValue` method to NOT trim whitespace? I think it could get a little weird, since this is written in such a way that the parent object keeps track of leading and trailing whitespace, so we'd have to either stick the value with leading and trailing whitespace into a concatenation node, make a new wrapper for simple config nodes that contains the whitespace, make `ConfigNodeSimpleValue` multi-token but not multi-node, or have a way to insert multiple nodes at once into a config node instead of just one.